src/db.c

```diff
@@ -1779,7 +1831,17 @@ robj *setExpire(client *c, serverDb *db, robj *key, long long when) {
     serverAssertWithInfo(NULL, key, valref != NULL);
     val = *valref;
     long long old_when = objectGetExpire(val);
+
+    /* If this is a hash with volatile fields, untrack it before setting the expire,
+     * since objectSetExpire may reallocate it. We'll re-track the new object after. */
+    bool updateHashExpiryKvsgtore = val->type == OBJ_HASH && val->encoding == OBJ_ENCODING_HASHTABLE && hashTypeHasVolatileElements(val);
+    if (updateHashExpiryKvsgtore) {
+        dbUntrackKeyWithVolaItems(db, val);
+    }
     robj *newval = objectSetExpire(val, when);
+    if (updateHashExpiryKvsgtore) {
+        dbTrackKeyWithVolaItems(db, newval);
+    }
```

Comment:
The code above will untrack/track the item even though most of the time, the item won't be reallocated.  I think this code will update the hashtable entry without accessing the old (deallocated) pointer.  If there are no weird side-effects in untrack/track, I think this should work.

But you might want to break out the guts into a separate function.


```suggestion
    robj *newval = objectSetExpire(val, when);
    if (updateHashExpiryKvsgtore && newval != val) {
            // Replace the pointer in the expire table without accessing the old pointer
            int dict_index = getKVStoreIndexForKey(objectGetKey(newval));
            hashtable *volatile_items_ht = kvstoreGetHashtable(db->keys_with_volatile_items, dict_index);
            int replaced = hashtableReplaceReallocatedEntry(volatile_items_ht, val, newval);
            serverAssert(replaced);
    }
```

Reply:
Thank you Jim! I think the issue was that we did not know hashtableReplaceReallocatedEntry exists